### ISSUE TITLE
Support some level of parallelization in junit-vintage-engine

### ISSUE THREAD

**Issue by sarod** [junit-team/junit5#2229]
## Goal
Add support for some level of parallelization inside the Junit Vintage engine.
Ideally the level of configuration should be similar to maven surefire parallel options.

## Why
When including junit 5 tests in a project containing junit4 tests the natural path is to use junit 5 platform with the jupiter and vintage engines.

However this causes problem with parallelization. 
In a project with Junit4 tests only maven surefire level parallelization works fine. (See https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html).
For Junit5 tests the jupiter engine [parallel execution works great](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-parallel-execution). 

However when a project contains a mix of Junit4 and Junit 5 there is no great way to configure parallelism:
* Surefire parallel does not work with Junit Platform Provider so it's not an option
* Surefire forkCount or gradle maxParallelForks can be used but:
  * It's not available at all when using  [tycho surefire](https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/plugin-info.html)
  * It introduces some memory and execution time overhead compared to surefire parallel or junit 5 parallel.
  * It provides less control than Junit 5 parallel execution options.
* Combining fork + junit 5 parallel execution leads to more parallel executions than expected  when executing junit 5 tests because parallelization is done at two level.

Another option is to configure two distinct test tasks one for junit 4 using surefire parallel and one for junit 5 but it kind of defeats the point of junit platform.

## Deliverables

- [x] Not sure

**Comment by koretzki** [comment]
thanks, @sarod for rising this. 
I tried to fully embrace Junit 5 but had issues working it using the maven failsafe plugin 2.22.2-3.0.0-M4 (Junit 5.3-5.6.2), but never could make parallelism to work even with settings that worked in maven surefire (in 3 ways to set them).
The issue was the mixed Junit 4,5 testing environment. I moved them all to Junit 5 and it worked like magic in maven failsafe 2.2.22 in parallel execution.

**Comment by vab2048** [comment]
Are there any known workarounds to achieve this currently? 
I would like to run the junit4 classes in parallel and methods in the same thread.

**Comment by marcphilipp** [comment]
>  I would like to run the junit4 classes in parallel and methods in the same thread.

That should be fairly straightforward to implement. Basically, one would have to call `RunnerExecutor` concurrently and read a few configuration parameters to enable parallel execution and configure the number of threads:
https://github.com/junit-team/junit5/blob/3f7fed61f2edcee2a5551dbde75382b1dbe21267/junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java#L77-L84

**Comment by kishoretak** [comment]
@marcphilipp As you suggested running test classes in parallel with Vintage Engine is straightforward and the same is working fine. Is there any suggestion for running the test methods from the same class in parallel? I tried the solution mentioned [here](https://github.com/junit-team/junit5/pull/2449) but it only supports test classes in parallel.

**Comment by nedtwigg** [comment]
I might be missing something, but so far as I can tell, it is not possible to run tests in parallel with junit-vintage. At the very least, `junit.jupiter.execution.parallel.enabled=true` does not seem to be effective. The PR #2449 looks small and high quality, I hope it or something similar gets merged someday.

If implementing the feature is very difficult, it might be nice if junit-vintage looked for `junit.jupiter.execution.parallel.enabled=true` and warned that it has no effect.

**Comment by nedtwigg** [comment]
For anybody else who wants this, I was able to get #2449 working on JitPack, took [some tricks](https://github.com/diffplug/l0s-junit5), you can use it like so:

```
maven { url 'https://jitpack.io' }

// testImplementation "org.junit.jupiter:junit-jupiter-api:${VER_junit5}"
// testImplementation "org.junit.vintage:junit-vintage-engine:${VER_junit5}"
testImplementation('com.github.diffplug.l0s-junit5:junit-jupiter-api:7786d92011e4c68255434b6f0f31cfd03841f24b')
testImplementation('com.github.diffplug.l0s-junit5:junit-vintage-engine:7786d92011e4c68255434b6f0f31cfd03841f24b')

test {
  useJUnitPlatform {}
  systemProperty 'junit.vintage.execution.parallel.enabled', 'true'
}
```

Swapping that little dep took a painpoint testsuite from 60 seconds to 10 seconds.

**Comment by YongGoose** [comment]
@marcphilipp 

I’m interested in this issue. 
If it’s still open for contributions, would it be okay for me to work on it?

**Comment by marcphilipp** [comment]
Yes, but let's discuss the plan first. What did you have in mind? Do sth. like I outlined in https://github.com/junit-team/junit5/issues/2229#issuecomment-691659555?

**Comment by YongGoose** [comment]
> Yes, but let's discuss the plan first. What did you have in mind? Do sth. like I outlined in [#2229 (comment)](https://github.com/junit-team/junit5/issues/2229#issuecomment-691659555)?

Yes, I think moving forward with the approach outlined in  [#2229 (comment)](https://github.com/junit-team/junit5/issues/2229#issuecomment-691659555) is a good idea.

I've implemented it in my personal workspace. Could you please check if the direction I took is correct? 
- https://github.com/YongGoose/junit5/commit/12d0a6e7c5035ee4d203e71c09f4bd2f40afdbaf

WDYT?

**Comment by YongGoose** [comment]
> I don't think using a ForkJoinPool as the Executor has any benefits here, does it?

Hmm... I think `ForkJoinPool` is more suitable for cases where tasks can be divided into **smaller units**. 
However, in the current code, since the tests are executed **independently**, there’s no way to split the tasks further.
That's why I understood there would be no benefit to using ForkJoinPool. 
- Did I understand it correctly? 🤔

I’ve changed it to use `newFixedThreadPool` as the Executor instead.

> Shouldn't we just wait for all futures (e.g. using CompletableFuture.allOf() rather than using a CountDownLatch?

Sounds Good!!, I think that's a good idea.

**Comment by marcphilipp** [comment]
> Did I understand it correctly?

Yes, you did.

> I've created a draft PR to make it easier to discuss details!

Thanks! 👍
